app/features/worksheet_generator/tools.py

```diff
@@ -0,0 +1,481 @@
+        except TypeError as e:
+            if self.verbose:
+                logger.error(f"TypeError during response validation: {e}")
+            return False
+    
+    def create_questions(self, num_questions: int = 5) -> List[Dict]:
+        if self.verbose: logger.info(f"Generating {num_questions} questions")
+        
+        if num_questions > 10:
+            return {"message": "error", "data": "Number of questions cannot exceed 10"}       
+        
+        generated_questions = []
+        attempts = 0
+        max_attempts = num_questions * 5  # Allow for more attempts to generate questions
+
+        while len(generated_questions) < num_questions and attempts < max_attempts:
+            response = self.generate_question()
+
+            if self.verbose:
+                logger.info(f"Generated response attempt {attempts + 1}: {response}")
+
+            if response is None:
+                if self.verbose:
+                    logger.warning(f"Invalid null response. Attempt {attempts + 1} of {max_attempts}")
+                attempts += 1
+                continue
+
+            #fix for the bug in Pydantic schema use, where the response contain properties key
+            if 'properties' in response:
+                # question generation is sucessful
+                if isinstance(response['properties']['model_config']['default']['json_schema_extra']['examples'], list):
+                    response = response['properties']['model_config']['default']['json_schema_extra']['examples'][0]
+                # question generation is unsuccessful
+                else:
+                    if self.verbose:
+                        logger.warning(f"Invalid response format. Attempt {attempts + 1} of {max_attempts}")
+                    attempts += 1
+                    continue
+
+            response = self.transform_json_dict(response)
+
+            # Directly check if the response format is valid
+            if self.validate_response(response):
+                generated_questions.append(response)
+                if self.verbose:
+                    logger.info(f"Valid question added: {response}")
+                    logger.info(f"Total generated questions: {len(generated_questions)}")
+            else:
+                if self.verbose:
+                    logger.warning(f"Invalid response format. Attempt {attempts + 1} of {max_attempts}")
+            
+            # Move to the next attempt regardless of success to ensure progress
+            attempts += 1
+
+        # Log if fewer questions are generated
+        if len(generated_questions) < num_questions:
+            logger.warning(f"Only generated {len(generated_questions)} out of {num_questions} requested questions")
+        
+        # Return the list of questions
+        return generated_questions[:num_questions]
+
+# Mapping of question types to their respective classes
+question_builder_classes = {
+    "multiple-choice": MultiChoiceQuestionBuilder,
```

Comment:
Check if you can divide your code in order to ensure a better readability and the Single Responsability principle